File: engine/backend/src/entity-handler.ts
import type { NatsConnection } from "nats";
import { StringCodec } from "nats";
import type { EntityId, ComponentId, ComponentSchema } from "@engine/core";
import { Subjects } from "@engine/core";
import { EntityRepository } from "./entity-repository.js";

const sc = StringCodec();

interface RegisteredComponent {
  readonly methodNames: string[];
  readonly schema: ComponentSchema;
}

export class EntityHandler {
  private readonly repo = new EntityRepository();
  private readonly components = new Map<string, RegisteredComponent>();

  constructor(private readonly nc: NatsConnection) {}

  async listen(): Promise<void> {
    this.handleRegisterComponent();
    this.handleListComponents();
    this.handleCreateEntity();
    this.handleDeleteEntity();
    this.handleHasEntity();
    this.handleListEntities();
    this.handleAddComponent();
    this.handleRemoveComponent();
    this.handleHasComponent();
    this.handleGetComponents();
    this.handleQueryEntity();
  }

  private handleRegisterComponent(): void {
    const sub = this.nc.subscribe(Subjects.registerComponent);
    (async () => {
      for await (const msg of sub) {
        try {
          const { componentId, methodNames, schema } = JSON.parse(
            sc.decode(msg.data),
          ) as {
            componentId: string;
            methodNames: string[];
            schema: ComponentSchema;
          };
          this.components.set(componentId, {
            methodNames,
            schema,
          });
          msg.respond(sc.encode(JSON.stringify({ ok: true })));
        } catch (e) {
          const message = e instanceof Error ? e.message : String(e);
          msg.respond(sc.encode(JSON.stringify({ error: message })));
        }
      }
    })();
  }

  private handleListComponents(): void {
    const sub = this.nc.subscribe(Subjects.listComponents);
    (async () => {
      for await (const msg of sub) {
        const entries = [...this.components.entries()].map(
          ([componentId, { methodNames, schema }]) => ({
            componentId,
            methodNames,
            schema,
          }),
        );
        msg.respond(sc.encode(JSON.stringify(entries)));
      }
    })();
  }

  private handleCreateEntity(): void {
    const sub = this.nc.subscribe(Subjects.createEntity);
    (async () => {
      for await (const msg of sub) {
        const id = this.repo.create();
        msg.respond(sc.encode(id));
      }
    })();
  }

  private handleDeleteEntity(): void {
    const sub = this.nc.subscribe(Subjects.deleteEntity);
    (async () => {
      for await (const msg of sub) {
        const id = sc.decode(msg.data) as EntityId;
        // Publish stop events for all workers on this entity
        const componentIds = this.repo.getComponentIds(id);
        for (const componentId of componentIds) {
          this.nc.publish(
            Subjects.stopWorker,
            sc.encode(JSON.stringify({ entityId: id, componentId })),
          );
        }
        const result = this.repo.delete(id);
        msg.respond(sc.encode(String(result)));
      }
    })();
  }

  private handleHasEntity(): void {
    const sub = this.nc.subscribe(Subjects.hasEntity);
    (async () => {
      for await (const msg of sub) {
        const id = sc.decode(msg.data) as EntityId;
        const result = this.repo.has(id);
        msg.respond(sc.encode(String(result)));
      }
    })();
  }

  private handleListEntities(): void {
    const sub = this.nc.subscribe(Subjects.listEntities);
    (async () => {
      for await (const msg of sub) {
        const ids = [...this.repo.getAll()];
        msg.respond(sc.encode(JSON.stringify(ids)));
      }
    })();
  }

  private handleAddComponent(): void {
    const sub = this.nc.subscribe(Subjects.addComponent);
    (async () => {
      for await (const msg of sub) {
        try {
          const { entityId, componentId } = JSON.parse(sc.decode(msg.data)) as {
            entityId: EntityId;
            componentId: ComponentId;
          };
          const registered = this.components.get(componentId as string);
          if (!registered) {
            msg.respond(
              sc.encode(
                JSON.stringify({
                  error: `No worker registered for component ${componentId as string}`,
                }),
              ),
            );
            continue;
          }

          this.repo.addComponent(entityId, componentId);

          // Publish start event — worker container will handle it
          this.nc.publish(
            Subjects.startWorker,
            sc.encode(JSON.stringify({ entityId, componentId })),
          );

          msg.respond(sc.encode(JSON.stringify({ ok: true })));
        } catch (e) {
          const message = e instanceof Error ? e.message : String(e);
          msg.respond(sc.encode(JSON.stringify({ error: message })));
        }
      }
    })();
  }

  private handleRemoveComponent(): void {
    const sub = this.nc.subscribe(Subjects.removeComponent);
    (async () => {
      for await (const msg of sub) {
        try {
          const { entityId, componentId } = JSON.parse(sc.decode(msg.data)) as {
            entityId: EntityId;
            componentId: ComponentId;
          };
          const result = this.repo.removeComponent(entityId, componentId);

          // Publish stop event — worker container will handle it
          this.nc.publish(
            Subjects.stopWorker,
            sc.encode(JSON.stringify({ entityId, componentId })),
          );

          msg.respond(sc.encode(String(result)));
        } catch (e) {
          const message = e instanceof Error ? e.message : String(e);
          msg.respond(sc.encode(JSON.stringify({ error: message })));
        }
      }
    })();
  }

  private handleHasComponent(): void {
    const sub = this.nc.subscribe(Subjects.hasComponent);
    (async () => {
      for await (const msg of sub) {
        const { entityId, componentId } = JSON.parse(sc.decode(msg.data)) as {
          entityId: EntityId;
          componentId: ComponentId;
        };
        const result = this.repo.hasComponent(entityId, componentId);
        msg.respond(sc.encode(String(result)));
      }
    })();
  }

  private handleGetComponents(): void {
    const sub = this.nc.subscribe(Subjects.getComponents);
    (async () => {
      for await (const msg of sub) {
        try {
          const entityId = sc.decode(msg.data) as EntityId;
          const componentIds = this.repo.getComponentIds(entityId);
          msg.respond(
            sc.encode(
              JSON.stringify(
                componentIds.map((id) => {
                  const registered = this.components.get(id as string);
                  return {
                    componentId: id as string,
                    methodNames: registered ? registered.methodNames : [],
                  };
                }),
              ),
            ),
          );
        } catch (e) {
          const message = e instanceof Error ? e.message : String(e);
          msg.respond(sc.encode(JSON.stringify({ error: message })));
        }
      }
    })();
  }

  private handleQueryEntity(): void {
    const sub = this.nc.subscribe(Subjects.queryEntity);
    (async () => {
      for await (const msg of sub) {
        try {
          const { entityId, methodNames } = JSON.parse(
            sc.decode(msg.data),
          ) as {
            entityId: EntityId;
            methodNames: string[];
          };

          const componentIds = this.repo.getComponentIds(entityId);

          // Build method → componentId mapping across all components on the entity
          const methodMap: Record<string, string> = {};
          for (const compId of componentIds) {
            const registered = this.components.get(compId as string);
            if (!registered) continue;
            for (const methodName of registered.methodNames) {
              // First component wins for a given method name
              if (!(methodName in methodMap)) {
                methodMap[methodName] = compId as string;
              }
            }
          }

          // Check if all requested methods are covered
          const missing = methodNames.filter((m) => !(m in methodMap));
          if (missing.length > 0) {
            msg.respond(
              sc.encode(JSON.stringify({ match: false, missing })),
            );
          } else {
            // Return only the requested methods' mapping
            const methods: Record<string, string> = {};
            for (const m of methodNames) {
              methods[m] = methodMap[m];
            }
            msg.respond(
              sc.encode(JSON.stringify({ match: true, methods })),
            );
          }
        } catch (e) {
          const message = e instanceof Error ? e.message : String(e);
          msg.respond(sc.encode(JSON.stringify({ error: message })));
        }
      }
    })();
  }
}

File: engine/backend/EntityHandler.cs
using System.Text.Json;
using Engine.Core;
using NATS.Client.Core;

namespace Engine.Backend;

public record RegisteredComponent(IReadOnlyList<string> MethodNames, JsonElement Schema);

public class EntityHandler
{
    private readonly EntityRepository _repo = new();
    private readonly Dictionary<string, RegisteredComponent> _components = new();
    private readonly INatsConnection _nc;
    private readonly CancellationToken _ct;

    public EntityHandler(INatsConnection nc, CancellationToken ct = default)
    {
        _nc = nc;
        _ct = ct;
    }

    public async Task ListenAsync()
    {
        var tasks = new[]
        {
            HandleRegisterComponent(),
            HandleListComponents(),
            HandleCreateEntity(),
            HandleDeleteEntity(),
            HandleHasEntity(),
            HandleListEntities(),
            HandleAddComponent(),
            HandleRemoveComponent(),
            HandleHasComponent(),
            HandleGetComponents(),
            HandleQueryEntity(),
        };

        await Task.WhenAll(tasks);
    }

    private async Task HandleRegisterComponent()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.RegisterComponent, cancellationToken: _ct))
        {
            try
            {
                using var doc = JsonDocument.Parse(msg.Data!);
                var root = doc.RootElement;
                var componentId = root.GetProperty("componentId").GetString()!;
                var methodNames = root.GetProperty("methodNames")
                    .EnumerateArray()
                    .Select(e => e.GetString()!)
                    .ToList();
                var schema = root.GetProperty("schema").Clone();

                _components[componentId] = new RegisteredComponent(methodNames, schema);
                await msg.ReplyAsync(JsonSerializer.Serialize(new { ok = true }), cancellationToken: _ct);
            }
            catch (Exception e)
            {
                await msg.ReplyAsync(JsonSerializer.Serialize(new { error = e.Message }), cancellationToken: _ct);
            }
        }
    }

    private async Task HandleListComponents()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.ListComponents, cancellationToken: _ct))
        {
            var entries = _components.Select(kvp => new
            {
                componentId = kvp.Key,
                methodNames = kvp.Value.MethodNames,
                schema = kvp.Value.Schema,
            });
            await msg.ReplyAsync(JsonSerializer.Serialize(entries), cancellationToken: _ct);
        }
    }

    private async Task HandleCreateEntity()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.CreateEntity, cancellationToken: _ct))
        {
            var id = _repo.Create();
            await msg.ReplyAsync(id.Value, cancellationToken: _ct);
        }
    }

    private async Task HandleDeleteEntity()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.DeleteEntity, cancellationToken: _ct))
        {
            var id = new EntityId(msg.Data!);
            var componentIds = _repo.GetComponentIds(id);
            foreach (var componentId in componentIds)
            {
                await _nc.PublishAsync(Subjects.StopWorker,
                    JsonSerializer.Serialize(new { entityId = id.Value, componentId }),
                    cancellationToken: _ct);
            }
            var result = _repo.Delete(id);
            await msg.ReplyAsync(result.ToString().ToLowerInvariant(), cancellationToken: _ct);
        }
    }

    private async Task HandleHasEntity()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.HasEntity, cancellationToken: _ct))
        {
            var id = new EntityId(msg.Data!);
            var result = _repo.Has(id);
            await msg.ReplyAsync(result.ToString().ToLowerInvariant(), cancellationToken: _ct);
        }
    }

    private async Task HandleListEntities()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.ListEntities, cancellationToken: _ct))
        {
            var ids = _repo.GetAll().Select(id => id.Value);
            await msg.ReplyAsync(JsonSerializer.Serialize(ids), cancellationToken: _ct);
        }
    }

    private async Task HandleAddComponent()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.AddComponent, cancellationToken: _ct))
        {
            try
            {
                using var doc = JsonDocument.Parse(msg.Data!);
                var root = doc.RootElement;
                var entityId = new EntityId(root.GetProperty("entityId").GetString()!);
                var componentId = root.GetProperty("componentId").GetString()!;

                if (!_components.ContainsKey(componentId))
                {
                    await msg.ReplyAsync(
                        JsonSerializer.Serialize(new { error = $"No worker registered for component {componentId}" }),
                        cancellationToken: _ct);
                    continue;
                }

                _repo.AddComponent(entityId, componentId);

                await _nc.PublishAsync(Subjects.StartWorker,
                    JsonSerializer.Serialize(new { entityId = entityId.Value, componentId }),
                    cancellationToken: _ct);

                await msg.ReplyAsync(JsonSerializer.Serialize(new { ok = true }), cancellationToken: _ct);
            }
            catch (Exception e)
            {
                await msg.ReplyAsync(JsonSerializer.Serialize(new { error = e.Message }), cancellationToken: _ct);
            }
        }
    }

    private async Task HandleRemoveComponent()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.RemoveComponent, cancellationToken: _ct))
        {
            try
            {
                using var doc = JsonDocument.Parse(msg.Data!);
                var root = doc.RootElement;
                var entityId = new EntityId(root.GetProperty("entityId").GetString()!);
                var componentId = root.GetProperty("componentId").GetString()!;

                var result = _repo.RemoveComponent(entityId, componentId);

                await _nc.PublishAsync(Subjects.StopWorker,
                    JsonSerializer.Serialize(new { entityId = entityId.Value, componentId }),
                    cancellationToken: _ct);

                await msg.ReplyAsync(result.ToString().ToLowerInvariant(), cancellationToken: _ct);
            }
            catch (Exception e)
            {
                await msg.ReplyAsync(JsonSerializer.Serialize(new { error = e.Message }), cancellationToken: _ct);
            }
        }
    }

    private async Task HandleHasComponent()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.HasComponent, cancellationToken: _ct))
        {
            using var doc = JsonDocument.Parse(msg.Data!);
            var root = doc.RootElement;
            var entityId = new EntityId(root.GetProperty("entityId").GetString()!);
            var componentId = root.GetProperty("componentId").GetString()!;

            var result = _repo.HasComponent(entityId, componentId);
            await msg.ReplyAsync(result.ToString().ToLowerInvariant(), cancellationToken: _ct);
        }
    }

    private async Task HandleGetComponents()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.GetComponents, cancellationToken: _ct))
        {
            try
            {
                var entityId = new EntityId(msg.Data!);
                var componentIds = _repo.GetComponentIds(entityId);
                var entries = componentIds.Select(id =>
                {
                    var methodNames = _components.TryGetValue(id, out var reg)
                        ? reg.MethodNames
                        : (IReadOnlyList<string>)Array.Empty<string>();
                    return new { componentId = id, methodNames };
                });
                await msg.ReplyAsync(JsonSerializer.Serialize(entries), cancellationToken: _ct);
            }
            catch (Exception e)
            {
                await msg.ReplyAsync(JsonSerializer.Serialize(new { error = e.Message }), cancellationToken: _ct);
            }
        }
    }

    private async Task HandleQueryEntity()
    {
        await foreach (var msg in _nc.SubscribeAsync<string>(Subjects.QueryEntity, cancellationToken: _ct))
        {
            try
            {
                using var doc = JsonDocument.Parse(msg.Data!);
                var root = doc.RootElement;
                var entityId = new EntityId(root.GetProperty("entityId").GetString()!);
                var methodNames = root.GetProperty("methodNames")
                    .EnumerateArray()
                    .Select(e => e.GetString()!)
                    .ToList();

                var componentIds = _repo.GetComponentIds(entityId);

                // Build method → componentId mapping across all components on entity
                var methodMap = new Dictionary<string, string>();
                foreach (var compId in componentIds)
                {
                    if (!_components.TryGetValue(compId, out var reg)) continue;
                    foreach (var methodName in reg.MethodNames)
                    {
                        // First component wins for a given method name
                        methodMap.TryAdd(methodName, compId);
                    }
                }

                var missing = methodNames.Where(m => !methodMap.ContainsKey(m)).ToList();
                if (missing.Count > 0)
                {
                    await msg.ReplyAsync(
                        JsonSerializer.Serialize(new { match = false, missing }),
                        cancellationToken: _ct);
                }
                else
                {
                    var methods = methodNames.ToDictionary(m => m, m => methodMap[m]);
                    await msg.ReplyAsync(
                        JsonSerializer.Serialize(new { match = true, methods }),
                        cancellationToken: _ct);
                }
            }
            catch (Exception e)
            {
                await msg.ReplyAsync(JsonSerializer.Serialize(new { error = e.Message }), cancellationToken: _ct);
            }
        }
    }
}
